test/dynamo/test_modes.py

```diff
@@ -35,6 +38,120 @@ def fn(x):
         self.assertEqual(cnt.frame_count, 0)
 
 
+class TorchFunctionModeTests(torch._dynamo.test_case.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+    @classmethod
+    def tearDownClass(cls):
+        super().tearDownClass()
+
+    def _run_torch_function_mode_guard_test(self):
+        class TestMode1(BaseTorchFunctionMode):
+            pass
+
+        class TestMode2(BaseTorchFunctionMode):
+            pass
+
+        cnt = torch._dynamo.testing.CompileCounter()
+
+        @torch.compile(backend=cnt.__call__)
+        def fn(x):
+            return x + 1
+
+        inp = torch.ones(2, 2)
+        fn(inp)
+        self.assertEqual(cnt.frame_count, 1)
+
+        with TestMode1():
+            fn(inp)
+        self.assertEqual(cnt.frame_count, 2)
+
+        with TestMode1(), TestMode2():
+            fn(inp)
+        self.assertEqual(cnt.frame_count, 3)
+
+        with TestMode2(), TestMode1():
+            fn(inp)
+        self.assertEqual(cnt.frame_count, 4)
+
+        with TestMode1():
+            fn(inp)
+        self.assertEqual(cnt.frame_count, 4)
+
+    def _run_ignored_mode_types_test(self):
+        class IgnoredMode(BaseTorchFunctionMode):
+            pass
+
+        cnt = torch._dynamo.testing.CompileCounter()
+
+        @torch.compile(backend=cnt.__call__, fullgraph=True)
+        def fn(x):
+            return x + 1
+
+        inp = torch.ones(2, 2)
+
+        with patch(
+            "torch._dynamo.variables.torch_function.IGNORED_MODES", {IgnoredMode}
```

Comment:
hmm. The `IGNORED_MODES` logic feels a bit like:

(1) it's sufficiently complicated that we need to add explicit testing for it to make sure we don't e.g. (here)

(2) it will all be removed once we properly treat `with torch.device()` like any other TorchFunctionMode in dynamo

(3) the `with device` context manager is also one of the more heavily-used `TorchFunctionMode`s we have today, so it would also provide some nice stress-testing for dynamo's mode handling.

Given that, it feels useful to try to treat `with device()` "normally" in dynamo as part of getting modes to work. WDYT?

Reply:
Yup this is accurate